feat: a hook to use Segment metrics

NicolasMassart opened the pull request:
## **Description**

- make MetaMetrics.getInstance synchronous and add an async init function + readiness check
- destructurable object hook return value to use metrics in components without having to use useRef and useEffect to set the instance
- interface for the destructurable object
- HOC (High Order Component) wrapper for class components
- new and updated unit tests

## **Related issues**

fixes https://github.com/MetaMask/mobile-planning/issues/1500

## **Manual testing steps**

none

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

NA

### **After**

NA

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

NicolasMassart committed:
metrics hook and HOC

and tests

wachunei commented on app/components/hooks/useMetrics/useMetrics.ts:
```diff
@@ -0,0 +1,65 @@
+function useMetrics() {
+  const [metrics, setMetrics] = useState<IMetaMetrics>();
+
+  useEffect(() => {
+    const setupMetaMetrics = async () => {
+      setMetrics(await MetaMetrics.getInstance());
+    };
+
+    setupMetaMetrics();
+  }, []);
```
How often are we expecting this to be unset when `trackEvent`, `trackAnonymousEvent` are called? I'm thinking this might be a very rare edge case.

NicolasMassart replied:
Do you mean we should add a readiness check?

NicolasMassart replied:
I changed the way we use metametrics so that's not an issue anymore

Cal-L commented on app/components/hooks/useMetrics/useMetrics.ts:
```diff
@@ -0,0 +1,318 @@
+
+  useEffect(() => {
+    const setupMetaMetrics = async () => {
+      setMetrics(await MetaMetrics.getInstance());
```
@NicolasMassart You mentioned that there's a risk to the instance not being defined before events are recorded and potential events being dropped. To optimize for this, we can expose an init method on the MetaMetrics class, which we can then trigger at an earlier point in the app. This should solve the initial concern while making it flexible enough for us to use depending when metrics is enabled or not.

Cal-L reviewed (commented):
Left some comments

NicolasMassart committed:
simplify hook

and add async init

wachunei commented on app/components/hooks/useMetrics/useMetrics.ts:
```diff
@@ -0,0 +1,121 @@
+  return {
+    trackEvent,
+    trackAnonymousEvent,
+    enable: MetaMetrics.getInstance().enable,
+    addTraitsToUser: MetaMetrics.getInstance().addTraitsToUser,
+    createDataDeletionTask: MetaMetrics.getInstance().createDataDeletionTask,
+    checkDataDeleteStatus: MetaMetrics.getInstance().checkDataDeleteStatus,
+    getDeleteRegulationCreationDate:
+      MetaMetrics.getInstance().getDeleteRegulationCreationDate,
+    getDeleteRegulationId: MetaMetrics.getInstance().getDeleteRegulationId,
+    isDataRecorded: MetaMetrics.getInstance().isDataRecorded,
+    isInitialized: MetaMetrics.getInstance().isInitialized,
+    isEnabled: MetaMetrics.getInstance().isEnabled,
+    init: MetaMetrics.getInstance().init,
```
I see there must be cases that some code needs to reach the MetaMetrics instance, and doing it by using the hook across the codebase is the right way, but this seems like it is too coupled? I'm thinking maybe we just expose `trackEvent, trackAnonymousEvent, getInstance` instead of remapping each instance method to the hook? wdyt?

NicolasMassart replied:
Great feedback!
Here's  what was my reasoning for choosing the mapping and not exposing the instance directly in the hook:
- I wanted to prevent the MetaMetrics class to embed component related code. So no effects or hooks or states or `InteractionManager.runAfterInteractions` for instance. Keeping it as pure as possible. Clean split between what we could call a controller and the more react native related side. 
- Hook is what makes the link with component context: adding lifecycle over metrics, for instance adding the `InteractionManager.runAfterInteractions` over trackEvent.
- If I expose the instance directly from the hook, in addition to wrapped function like trackEvent, it means devs will have the ability use the instance trackEvent. The later being dangerous as it's not taking the lifecycle specific behaviour.
- I wanted to expose only a subset of features that user will really use. Group, Reset and Alias for instance are not useful for the moment, so not exposing them seems the thing to do. They are usable through the MetaMetrics instance directly though. MetaMetrics class implements the full Segment SDK logic.
- The hook makes it far easier to test components and spy/mock the functions as we have some component tests that check the events are tracked. Which is a good thing we should do more.
- for advanced usage, we could directly use the MetaMetrics instance without going through the hook.
- hook greatly simplifies the event tracking calls migration, and given the amount of calls we have, and it can not really be done with search/replace, that will help speeding the process.

I know, probably it would be fine to directly forward instance. Maybe. This is just the thoughts I took onto account for my choice. Note that this is evolving.
I would be happy to update if that's really not ideal. In terms of maintainability, I believe this is not a big deal as the functions are tied to Segment SDK and I don't thing we will have new features often.
Anyway, even if it's the case, the mapping is still valid unless the signatures change. Code is tested and documented, so any change should be easy and not so impactful.

Cal-L commented on app/components/hooks/useMetrics/withMetricsAwareness.tsx:
```diff
@@ -0,0 +1,16 @@
+const withMetricsAwareness =
+  (
+    Children: ComponentClass<{
+      metrics: IUseMetricsHook;
```
Let's make this a type as well so we can use it here and in other components when using withMetricsAwareness

Cal-L reviewed (commented):
Left some comments

NicolasMassart commented on app/components/hooks/useMetrics/withMetricsAwareness.tsx:
```diff
@@ -0,0 +1,16 @@
+const withMetricsAwareness =
+  (
+    Children: ComponentClass<{
+      metrics: IUseMetricsHook;
```
Could you clarify what you would want me to make a type with? Could you suggest a change using the ````suggestion` comment ?

NicolasMassart commented on app/core/Analytics/MetaMetrics.ts:
```diff
@@ -462,19 +461,39 @@ class MetaMetrics implements IMetaMetrics {
+   * @returns Promise indicating if MetaMetrics was initialized or not
+   */
+  init = async (): Promise<boolean> => {
+    if (this.#isInitialized) return true;
```
What do you mean? using Instance doesn't tell if it's initialized or not. Instance is defined as soon as you create it. But init may not have been async called. Could you show me a suggested change please?

NicolasMassart commented on app/core/Analytics/MetaMetrics.ts:
```diff
@@ -462,19 +461,39 @@ class MetaMetrics implements IMetaMetrics {
+   * @returns Promise indicating if MetaMetrics was initialized or not
+   */
+  init = async (): Promise<boolean> => {
+    if (this.#isInitialized) return true;
```
Returning early is to prevent useless computation in case it's already initialised. Do you mean something different?

NicolasMassart committed:
rename init to configure and remove it from hook

- use interface for props types
- keep isConfigured private only
- update tests
- update HOC

NicolasMassart committed:
update doc

Cal-L commented on app/components/hooks/useMetrics/useMetrics.test.tsx:
```diff
@@ -0,0 +1,146 @@
+  isDataRecorded: jest.fn(() => true),
+  isInitialized: jest.fn(() => false),
+  isEnabled: jest.fn(() => true),
+  init: jest.fn(() => Promise.resolve(false)),
```
Remove init and isInitialized from tests

NicolasMassart committed:
remove unused mock func

Cal-L reviewed (approved):
LGTM
